lib/Sema/TypeCheckCaptures.cpp

```diff
@@ -568,7 +560,7 @@ class FindCapturedVars : public ASTWalker {
 
     // Capture a placeholder opaque value.
     if (auto opaqueValue = dyn_cast<OpaqueValueExpr>(E)) {
-      if (opaqueValue->isPlaceholder()) {
+      if (opaqueValue->getOwnerKind() == ExprKind::Call /* FIXME */) {
```

Comment:
It looks like this is the only time that getOwnerKind() is checked outside of an assertion. Perhaps the capture analysis can track which Exprs introduce which OVEs, and then record and "free" OVEs, instead?

Reply:
Having gone through the exercise of doing this, there seem to currently be two interesting pieces of information: (1) whether this OVE can be captured and (2) whether we should ever see this OVE before CSApply. I could reduce it down to those two bits instead of storing the whole ExprKind (which doesn't quite work anyway).

Alternatively, I think the only problem with capturing OVEs is that Sema only knows how to analyzes a single OVE capture, even though we could represent arbitrary numbers of them as `CaptureValues`. If that's true, we could refactor to allow *n* OVE captures.

(One thing I noticed while working on this is that there are several different walks in the compiler which each separately have knowledge of which nodes can own OVEs. Is there some way we could centralize that knowledge? Maybe an `ASTWalker` subclass that automatically maintains OVE -> owner tables?)